fiftyone/plugins/core.py

```diff
@@ -49,498 +48,473 @@ def __repr__(self):
+            same name if it already exists
+
+    Returns:
+        the path to the downloaded plugin
+    """
+    if etaw.is_url(url_or_gh_repo):
+        if "github" in url_or_gh_repo:
+            repo = GitHubRepository(url)
+            url = repo.download_url
+        else:
+            url = url_or_gh_repo
+    else:
+        repo = GitHubRepository(url_or_gh_repo)
+        url = repo.download_url
+
+    return _download_plugin(
+        url,
+        plugin_names=plugin_names,
+        max_depth=max_depth,
+        overwrite=overwrite,
     )
 
 
-def _cleanup_plugin_dir(plugin_dir: str, recursive: bool = True):
-    """Deletes any empty plugin parent directories if empty after deleting the plugin."""
-    if plugin_dir in _PLUGIN_DIRS:
-        return
-    plugin_parent_dir = os.path.dirname(plugin_dir)
-    if len(glob.glob(os.path.join(plugin_parent_dir, "*"))) == 0:
-        if plugin_parent_dir not in _PLUGIN_DIRS:
-            shutil.rmtree(plugin_parent_dir)
-        logging.debug(f"Deleted empty directory {plugin_parent_dir}")
-    if recursive:
-        return _cleanup_plugin_dir(plugin_parent_dir, recursive=recursive)
+def _download_plugin(url, plugin_names=None, max_depth=3, overwrite=False):
+    if etau.is_str(plugin_names):
+        plugin_names = {plugin_names}
+    elif plugin_names is not None:
+        plugin_names = set(plugin_names)
 
+    existing_plugin_dirs = {p.name: p.path for p in _list_plugins()}
 
-def list_downloaded_plugins():
-    """Returns a list of all downloaded plugins by name."""
-    return _list_plugins_by_name()
+    with etau.TempDir() as tmpdir:
+        path = os.path.join(tmpdir, os.path.basename(url))
+        etaw.download_file(url, path=path)
+        etau.extract_archive(path)
 
+        metadata_paths = _find_plugin_metadata_files(
+            os.path.join(tmpdir, "*"),
+            max_depth=max_depth + 1,
+        )
 
-def list_enabled_plugins():
-    """Returns a list of all enabled plugins by name."""
-    return _list_plugins_by_name(enabled_only=True)
+        for metadata_path in metadata_paths:
+            try:
+                plugin = _parse_plugin_metadata(metadata_path)
+            except AttributeError:
+                logging.debug(f"Failed to parse '{metadata_path}'")
+                continue
+
+            if plugin_names is not None and plugin.name not in plugin_names:
+                logger.debug(f"Skipping unwanted plugin '{plugin.name}'")
+                continue
+
+            existing_dir = existing_plugin_dirs.get(plugin.name, None)
+            if existing_dir is not None:
+                if not overwrite:
+                    logger.debug(f"Skipping existing plugin '{plugin.name}'")
+                    continue
 
+                logger.debug(f"Overwriting existing plugin '{plugin.name}'")
+                etau.delete_dir(existing_dir)
+                plugin_dir = existing_dir
+            else:
+                plugin_dir = _recommend_plugin_dir(plugin)
```

Comment:
Can we keep it as before and use the url as the directory name? This helps separate locally developed and downloaded under most cases, enables checking if the archive needed to be redownloaded, and also keeps track of download source

Reply:
It's complicated because if the user redownloads and `overwrite=True`, the current logic will update the existing directory in-place. If we keep the download hash in the folder tree in `PLUGINS_DIR`, that will be confusing because the new version may not be from that hash.

Reply:
An alternative approach would be to inject some metadata within the downloaded directory that notes when the download happened, where it came from, etc.